### ISSUE TITLE
F.gumbel_softmax returns NaN on MPS device

### ISSUE THREAD

**Issue by ageron** [pytorch/pytorch#159103]
### 🐛 Describe the bug

Running this code on a MacBook with MPS acceleration, the assertion fails (in my case when `seed==28`):

```python
import torch
import torch.nn.functional as F

for seed in range(1000):
    torch.manual_seed(seed)
    logits = torch.zeros(32768, device="mps")
    gumbels = F.gumbel_softmax(logits)
    assert not torch.isnan(gumbels).any()
```

If you replace `device="mps"` with `device="cpu"` (or with `device="cuda"` on a CUDA machine) then everything works fine.

The culprit seems to be the noise generation process in the `F.gumbel_softmax()` function, which sometimes generates infinite values (which get turned into NaNs by the `torch.softmax()` function. The following code bombs on MPS devices but not on CPU or CUDA:

```python
import torch

for seed in range(1000):
    torch.manual_seed(seed)
    gumbels = torch.empty(32768, device="mps").exponential_().log()
    assert not torch.isinf(gumbels).any()
```

The reason seems to be that the `exponential_()` method sometimes produces actual zeros, which the `log()` method turns into infinities.

Maybe similar to #2561 ?

As a workaround, I've copied the function and clamped the gumbels to the -30 to +30 range, and everything worked fine after that.

```python
def gumbel_softmax(logits, tau=1, hard=False, dim=-1):
    gumbels = (
        -torch.empty_like(logits, memory_format=torch.legacy_contiguous_format)
        .exponential_()
        .log()
    )  # ~Gumbel(0,1)
    gumbels = torch.clamp(gumbels, -30, 30)  # <<<<<<<<<<<<<<<<<<<<< ADDED
    gumbels = (logits + gumbels) / tau  # ~Gumbel(logits,tau)
    y_soft = gumbels.softmax(dim)

    if hard:
        # Straight through.
        index = y_soft.max(dim, keepdim=True)[1]
        y_hard = torch.zeros_like(
            logits, memory_format=torch.legacy_contiguous_format
        ).scatter_(dim, index, 1.0)
        ret = y_hard - y_soft.detach() + y_soft
    else:
        # Reparametrization trick.
        ret = y_soft
    return ret
```

### Versions

Collecting environment information...
PyTorch version: 2.6.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.5 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: version 4.0.1
Libc version: N/A

Python version: 3.12.10 | packaged by conda-forge | (main, Apr 10 2025, 22:19:24) [Clang 18.1.8 ] (64-bit runtime)
Python platform: macOS-15.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M4 Max

Versions of relevant libraries:
[pip3] lion-pytorch==0.2.3
[pip3] numpy==1.26.4
[pip3] optree==0.15.0
[pip3] torch==2.6.0
[pip3] torch-tb-profiler==0.4.3
[pip3] torchaudio==2.7.0
[pip3] torchmetrics==1.7.1
[pip3] torchvision==0.21.0
[pip3] torchvision-extra-decoders==0.0.2
[conda] Could not collect

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by kurtamohler** [comment]
This looks like a real issue to me, and I'm able to reproduce it. The implementation of `gumbel_softmax` calls `exponential_`, and `exponential_` does indeed output zeros on MPS. On CPU, `exponential_` doesn't output zeros. The [doc for `exponential_` ](https://docs.pytorch.org/docs/stable/generated/torch.Tensor.exponential_.html#torch.Tensor.exponential_) says that the function is only supposed to output values above 0. I can submit a patch for this.
